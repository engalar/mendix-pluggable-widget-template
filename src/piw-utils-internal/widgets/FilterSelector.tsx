import { createElement, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import classNames from "classnames";
import { useOnClickOutside } from "../utils";

interface FilterSelectorProps<T> {
    ariaLabel?: string;
    name?: string;
    defaultFilter: T;
    onChange: (value: T) => void;
    options: Array<{ value: T; label: string }>;
}

export function FilterSelector<T>(props: FilterSelectorProps<T>): ReactElement {
    const [value, setValue] = useState(props.defaultFilter);
    const [show, setShow] = useState(false);
    const componentRef = useRef<HTMLDivElement>(null);
    useOnClickOutside(componentRef, () => setShow(false));

    const onClick = useCallback(
        (value: T) => {
            setValue(value);
            props.onChange(value);
            setShow(false);
        },
        [props.onChange]
    );

    useEffect(() => {
        setValue(props.defaultFilter);
        props.onChange(props.defaultFilter);
    }, [props.defaultFilter, props.onChange]);

    return (
        <div className="filter-selector">
            <div className="filter-selector-content" ref={componentRef}>
                <button
                    aria-controls={`${props.name}-filter-selectors`}
                    aria-expanded={show}
                    aria-haspopup
                    aria-label={props.ariaLabel}
                    className={classNames("btn btn-default filter-selector-button button-icon", value)}
                    onClick={() => setShow(show => !show)}
                >
                    &nbsp;
                </button>
                {show && (
                    <ul
                        id={`${props.name}-filter-selectors`}
                        className="filter-selectors"
                        role="menu"
                        data-focusindex={0}
                    >
                        {props.options.map((option, index) => (
                            <li
                                className={classNames({ "filter-selected": value === option.value })}
                                key={index}
                                onClick={() => onClick(option.value)}
                                onKeyDown={e => {
                                    if (e.key === "Enter" || e.key === " ") {
                                        e.preventDefault();
                                        onClick(option.value);
                                    }
                                }}
                                role="menuitem"
                                tabIndex={0}
                            >
                                <div className={classNames("filter-icon", option.value)} aria-hidden />
                                <div className="filter-label">{option.label}</div>
                            </li>
                        ))}
                    </ul>
                )}
            </div>
        </div>
    );
}
