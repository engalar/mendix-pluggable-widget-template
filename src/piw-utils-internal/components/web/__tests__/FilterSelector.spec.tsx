import { shallow } from "enzyme";
import { render, fireEvent, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { FilterSelector } from "../FilterSelector";

const options = [
    { value: "contains", label: "Contains" },
    { value: "startsWith", label: "Starts with" },
    { value: "endsWith", label: "Ends with" },
    { value: "greater", label: "Greater than" },
    { value: "greaterEqual", label: "Greater than or equal" },
    { value: "equal", label: "Equal" },
    { value: "notEqual", label: "Not equal" },
    { value: "smaller", label: "Smaller than" },
    { value: "smallerEqual", label: "Smaller than or equal" }
];

jest.useFakeTimers();

describe("Filter selector", () => {
    it("renders correctly", () => {
        const component = shallow(
            <FilterSelector defaultFilter="contains" onChange={jest.fn()} id="test" options={options} />
        );

        expect(component).toMatchSnapshot();
    });

    it("renders correctly with filter selectors open", () => {
        const onClickProps = { preventDefault: jest.fn(), stopPropagation: jest.fn() };
        const onChange = jest.fn();
        const component = shallow(
            <FilterSelector defaultFilter="contains" onChange={onChange} id="test" options={options} />
        );

        const button = component.find("button");
        button.simulate("click", onClickProps);

        expect(component).toMatchSnapshot();
    });

    it("renders correctly with aria-label", () => {
        const component = shallow(
            <FilterSelector
                ariaLabel="my label"
                defaultFilter="contains"
                onChange={jest.fn()}
                id="test"
                options={options}
            />
        );

        expect(component).toMatchSnapshot();
    });

    it("renders correctly with another default filter", () => {
        const component = shallow(
            <FilterSelector defaultFilter="equal" onChange={jest.fn()} id="test" options={options} />
        );

        expect(component).toMatchSnapshot();
    });

    it("calls onChange when type changes", () => {
        const onClickProps = { preventDefault: jest.fn(), stopPropagation: jest.fn() };
        const onChange = jest.fn();
        const component = shallow(
            <FilterSelector defaultFilter="contains" onChange={onChange} id="test" options={options} />
        );

        const button = component.find("button");
        button.simulate("click", onClickProps);
        const lis = component.find("li");

        expect(lis.at(0)).toBeDefined();
        lis.at(0).simulate("click", onClickProps);

        expect(onChange).toBeCalled();
        expect(onChange).toBeCalledWith("contains");

        lis.at(1).simulate("click", onClickProps);
        expect(onChange).toBeCalledWith("startsWith");
    });

    describe("focus", () => {
        beforeEach(() => (document.body.innerHTML = ""));

        it("changes focused element when pressing filter selector button", () => {
            render(<FilterSelector defaultFilter="contains" onChange={jest.fn()} id="test" options={options} />);

            expect(document.body).toHaveFocus();

            const button = screen.getByRole("button");
            expect(button).toBeDefined();
            fireEvent.click(button);

            jest.advanceTimersByTime(10);

            const items = screen.getAllByRole("menuitem");

            expect(items[0]).toHaveFocus();
        });

        it("changes focused element back to the button when pressing shift+tab in the first element", () => {
            render(<FilterSelector defaultFilter="contains" onChange={jest.fn()} id="test" options={options} />);

            expect(document.body).toHaveFocus();

            const button = screen.getByRole("button");
            expect(button).toBeDefined();
            fireEvent.click(button);

            jest.advanceTimersByTime(10);

            const items = screen.getAllByRole("menuitem");
            expect(items[0]).toHaveFocus();

            userEvent.tab({ shift: true });

            jest.advanceTimersByTime(10);

            expect(button).toHaveFocus();
        });

        it("triggers onChange with previous value when pressing tab on the last item", () => {
            const onChange = jest.fn();

            render(
                <FilterSelector
                    defaultFilter="contains"
                    onChange={onChange}
                    id="test"
                    options={[
                        { value: "contains", label: "Contains" },
                        { value: "startsWith", label: "Starts with" }
                    ]}
                />
            );

            expect(document.body).toHaveFocus();

            const button = screen.getByRole("button");
            expect(button).toBeDefined();
            fireEvent.click(button);

            jest.advanceTimersByTime(10);

            const items = screen.getAllByRole("menuitem");
            expect(items[0]).toHaveFocus();

            userEvent.tab();
            expect(items[1]).toHaveFocus();
            userEvent.tab();

            jest.advanceTimersByTime(10);

            expect(onChange).toHaveBeenCalledWith("contains");
        });

        it("changes focused element back to the button when pressing escape in any element", () => {
            render(<FilterSelector defaultFilter="contains" onChange={jest.fn()} id="test" options={options} />);

            expect(document.body).toHaveFocus();

            const button = screen.getByRole("button");
            expect(button).toBeDefined();
            fireEvent.click(button);

            jest.advanceTimersByTime(10);

            const items = screen.getAllByRole("menuitem");
            expect(items[0]).toHaveFocus();

            userEvent.tab();

            expect(items[1]).toHaveFocus();

            userEvent.keyboard("{esc}");

            jest.advanceTimersByTime(10);

            expect(button).toHaveFocus();
        });
    });
});
