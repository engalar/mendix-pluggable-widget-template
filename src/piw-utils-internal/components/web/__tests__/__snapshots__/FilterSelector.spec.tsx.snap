// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter selector renders correctly 1`] = `
<div
  className="filter-selector"
>
  <div
    className="filter-selector-content"
  >
    <button
      aria-controls="test-filter-selectors"
      aria-expanded={false}
      aria-haspopup={true}
      className="btn btn-default filter-selector-button button-icon contains"
      onClick={[Function]}
      onKeyDown={[Function]}
    >
       
    </button>
  </div>
</div>
`;

exports[`Filter selector renders correctly with another default filter 1`] = `
<div
  className="filter-selector"
>
  <div
    className="filter-selector-content"
  >
    <button
      aria-controls="test-filter-selectors"
      aria-expanded={false}
      aria-haspopup={true}
      className="btn btn-default filter-selector-button button-icon equal"
      onClick={[Function]}
      onKeyDown={[Function]}
    >
       
    </button>
  </div>
</div>
`;

exports[`Filter selector renders correctly with aria-label 1`] = `
<div
  className="filter-selector"
>
  <div
    className="filter-selector-content"
  >
    <button
      aria-controls="test-filter-selectors"
      aria-expanded={false}
      aria-haspopup={true}
      aria-label="my label"
      className="btn btn-default filter-selector-button button-icon contains"
      onClick={[Function]}
      onKeyDown={[Function]}
    >
       
    </button>
  </div>
</div>
`;

exports[`Filter selector renders correctly with filter selectors open 1`] = `
<div
  className="filter-selector"
>
  <div
    className="filter-selector-content"
  >
    <button
      aria-controls="test-filter-selectors"
      aria-expanded={true}
      aria-haspopup={true}
      className="btn btn-default filter-selector-button button-icon contains"
      onClick={[Function]}
      onKeyDown={[Function]}
    >
       
    </button>
    <Portal
      containerInfo={<body />}
    >
      <ul
        className="filter-selectors"
        data-focusindex={0}
        id="test-filter-selectors"
        role="menu"
        style={
          Object {
            "left": undefined,
            "position": "fixed",
            "top": undefined,
          }
        }
      >
        <li
          className="filter-selected"
          key="0"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex={0}
        >
          <div
            aria-hidden={true}
            className="filter-icon contains"
          />
          <div
            className="filter-label"
          >
            Contains
          </div>
        </li>
        <li
          className=""
          key="1"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex={0}
        >
          <div
            aria-hidden={true}
            className="filter-icon startsWith"
          />
          <div
            className="filter-label"
          >
            Starts with
          </div>
        </li>
        <li
          className=""
          key="2"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex={0}
        >
          <div
            aria-hidden={true}
            className="filter-icon endsWith"
          />
          <div
            className="filter-label"
          >
            Ends with
          </div>
        </li>
        <li
          className=""
          key="3"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex={0}
        >
          <div
            aria-hidden={true}
            className="filter-icon greater"
          />
          <div
            className="filter-label"
          >
            Greater than
          </div>
        </li>
        <li
          className=""
          key="4"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex={0}
        >
          <div
            aria-hidden={true}
            className="filter-icon greaterEqual"
          />
          <div
            className="filter-label"
          >
            Greater than or equal
          </div>
        </li>
        <li
          className=""
          key="5"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex={0}
        >
          <div
            aria-hidden={true}
            className="filter-icon equal"
          />
          <div
            className="filter-label"
          >
            Equal
          </div>
        </li>
        <li
          className=""
          key="6"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex={0}
        >
          <div
            aria-hidden={true}
            className="filter-icon notEqual"
          />
          <div
            className="filter-label"
          >
            Not equal
          </div>
        </li>
        <li
          className=""
          key="7"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex={0}
        >
          <div
            aria-hidden={true}
            className="filter-icon smaller"
          />
          <div
            className="filter-label"
          >
            Smaller than
          </div>
        </li>
        <li
          className=""
          key="8"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex={0}
        >
          <div
            aria-hidden={true}
            className="filter-icon smallerEqual"
          />
          <div
            className="filter-label"
          >
            Smaller than or equal
          </div>
        </li>
      </ul>
    </Portal>
  </div>
</div>
`;
